src/walletConnectV2Provider.ts

```diff
@@ -832,19 +765,58 @@ export class WalletConnectV2Provider {
       const session = this.getCurrentSession(client);
       if (session) {
         await this.onSessionConnected({ session });
+
         return session;
       }
     }
 
     return;
   }
 
+  private async cleanupPendingPairings(
+    opts: { deletePairings?: boolean } = {}
+  ): Promise<void> {
+    if (typeof this.walletConnector === "undefined") {
+      return;
+    }
+
+    try {
+      const inactivePairings =
+        this.walletConnector?.core?.pairing?.pairings.getAll();
+
+      if (!isValidArray(inactivePairings)) return;
+
+      for (const pairing of inactivePairings) {
+        if (opts?.deletePairings) {
+          this.walletConnector?.core?.expirer?.set(pairing.topic, 0);
+        } else {
+          await this.walletConnector?.core?.relayer?.subscriber?.unsubscribe(
+            pairing.topic
+          );
+        }
+      }
+    } catch (error) {
+      Logger.error(
+        WalletConnectV2ProviderErrorMessagesEnum.unableToHandleCleanup
+      );
+    }
+  }
+
   private getCurrentSession(client: Client): SessionTypes.Struct {
     if (typeof client === "undefined") {
       throw new Error(WalletConnectV2ProviderErrorMessagesEnum.notInitialized);
     }
 
-    if (client.session.length) {
+    const sessions = client
+      .find(this.getConnectionParams())
+      .filter((s) => s.acknowledged);
+
+    if (sessions.length > 0) {
+      const lastKeyIndex = sessions.length - 1;
+      const session = sessions[lastKeyIndex];
+
+      return session;
+    } else if (client.session.length > 0) {
```

Comment:
Renamed as suggested to acknowledgedSessions
The second case is in case the WalletConnect engine logic changes. Shouldn't happen, but it's better to be safe